presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java

```diff
@@ -229,8 +225,22 @@ public void finish()
                 return;
             }
 
+            long size = 0;
+            LookupSource[] lookupSources = new LookupSource[partitions.length];
             for (int partition = 0; partition < partitions.length; partition++) {
-                partitionFutures.get(partition).set(partitions[partition]);
+                LookupSource source = partitions[partition].createLookupSource(hashChannels, hashChannel);
+                size += source.getInMemorySizeInBytes();
+                lookupSources[partition] = source;
+            }
+            // After this point the SharedLookupSources will take over our memory reservation, and ours will be zero
+            operatorContext.transferMemoryToTaskContext(size);
```

Comment:
I had to add a guard `OperatorContext.transferMemoryToTaskContext` called only once.